### PR TITLE
Segmentation fault in reduction with logging redirection on Mantid 6.0

### PR DESCRIPTION
**Pull request by jmborr**
Companion to PR #31861  
**Description of issue**  
See [SANS771](https://code.ornl.gov/sns-hfir-scse/sans/sans-backend/-/issues/771)

**Description of work.**  
- [x] Acquire GIL for the `PythonStdoutChannel` class
- [x] unit test
- [x] No release notes necessary because these changes don't change the public API

**To test:**
open the workbench and run the script(s) of your choice, inserting the following lines at the very beginning:
```python
from mantid.kernel import ConfigService
ConfigService.setString('logging.channels.consoleChannel.class', 'PythonStdoutChannel')
```
The script(s) should run as usual

*There is no associated issue.*

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/PythonInterface/core/inc/MantidPythonInterface/core/PythonStdoutChannel.h [hunk 1]
```diff
@@ -27,4 +37,17 @@ class MANTID_PYTHONINTERFACE_CORE_DLL PythonStdoutChannel : public ConsoleChanne
...
+
+/* /// Store in case class PythonStdoutChannel is not up to the task and needs to be replaced with this class
+class MANTID_PYTHONINTERFACE_CORE_DLL PyBindStdoutChannel : public StdoutChannel {
+
+public:
+  /// Constructor for PythonStdoutChannel
+  PyBindStdoutChannel() : StdoutChannel() {}
+
+private:
+  pybind11::scoped_ostream_redirect m_redirect;
```
Comment (wdzhou) [inline Framework/PythonInterface/core/inc/MantidPythonInterface/core/PythonStdoutChannel.h:49]: If the compiler is C++17 or later, will this `pybind11` work still? 

This is the list of compilers that it supports now.
```
Supported compilers
Clang/LLVM 3.3 or newer (for Apple Xcode’s clang, this is 5.0.0 or newer)
GCC 4.8 or newer
Microsoft Visual Studio 2015 Update 3 or newer
Intel classic C++ compiler 18 or newer (ICC 20.2 tested in CI)
Cygwin/GCC (previously tested on 2.5.1)
NVCC (CUDA 11.0 tested in CI)
NVIDIA PGI (20.9 tested in CI)
```

File: Framework/PythonInterface/core/src/PythonStdoutChannel.cpp [hunk 2]
```diff
@@ -14,28 +14,30 @@
...
 // wrapper of that sink to be a stream
 PyStdoutSink pyStdoutSinkInstance = PyStdoutSink(); // needs to be initialized separately
-boost::iostreams::stream<PyStdoutSink> PyStdout(pyStdoutSinkInstance);
+boost::iostreams::stream<PyStdoutSink> PyStdoutStream(pyStdoutSinkInstance);
 
 } // anonymous namespace
 
 namespace Poco {
-PythonStdoutChannel::PythonStdoutChannel() : ConsoleChannel(PyStdout) {}
+PythonStdoutChannel::PythonStdoutChannel() : ConsoleChannel(PyStdoutStream) {}
```
Comment (wdzhou) [inline Framework/PythonInterface/core/src/PythonStdoutChannel.cpp:42]: This sounds a "fix" to a bug.

File: Framework/PythonInterface/mantid/kernel/src/kernel.cpp.in [hunk 3]
```diff
@@ -51,6 +51,10 @@ BOOST_PYTHON_MODULE(_kernel) {
   Poco::LoggingFactory::defaultFactory().registerChannelClass(
     "PythonStdoutChannel",
     new Poco::Instantiator<Poco::PythonStdoutChannel, Poco::Channel>);
+  /* // Store in case class PythonStdoutChannel is not up to the task and needs to be replaced with this class
```
Comment (wdzhou) [inline Framework/PythonInterface/mantid/kernel/src/kernel.cpp.in:54]: A comment?  Applied anywhere?

File: Framework/PythonInterface/core/inc/MantidPythonInterface/core/PythonStdoutChannel.h [hunk 4]
```diff
@@ -17,8 +17,18 @@
...
 
+// local includes
 #include "MantidPythonInterface/core/DllConfig.h"
+#include "MantidPythonInterface/core/GlobalInterpreterLock.h"
+#include "MantidPythonInterface/core/WrapPython.h"
+
+// 3rd-party includes
+#include "MantidKernel/StdoutChannel.h"
 #include <Poco/ConsoleChannel.h>
+//#include <pybind11/iostream.h>
```
Comment (wdzhou) [inline Framework/PythonInterface/core/inc/MantidPythonInterface/core/PythonStdoutChannel.h:28]: delete?

File: Framework/PythonInterface/core/inc/MantidPythonInterface/core/PythonStdoutChannel.h (hunk 1 above)
Comment (wdzhou) [reply Framework/PythonInterface/core/inc/MantidPythonInterface/core/PythonStdoutChannel.h:49]: But it seems that you "deleted" it already.

File: Framework/PythonInterface/core/src/PythonStdoutChannel.cpp [hunk 5]
```diff
@@ -14,28 +14,30 @@
...
+    Mantid::PythonInterface::GlobalInterpreterLock gil; // acquire the GIL
     // PySys_WriteStdout truncates to 1000 chars
     static const std::streamsize MAXSIZE = 1000;
 
     std::streamsize written = std::min(n, MAXSIZE);
     PySys_WriteStdout((boost::format("%%.%1%s") % written).str().c_str(), s);
 
     return written;
-  }
+  } // release the GIL
```
Comment (wdzhou) [inline Framework/PythonInterface/core/src/PythonStdoutChannel.cpp:32]: This seems to be a "fix".

**Review by wdzhou — COMMENTED** [review]
Overall, some commented out codes shall be deleted.

File: Framework/PythonInterface/core/src/PythonStdoutChannel.cpp (hunk 5 above)
Comment (jmborr) [reply Framework/PythonInterface/core/src/PythonStdoutChannel.cpp:32]: It is a fix.
I wrote a new new python channel but removed it later, seeing that acquiring the GIL fixed this channel

File: Framework/PythonInterface/core/inc/MantidPythonInterface/core/PythonStdoutChannel.h (hunk 1 above)
Comment (jmborr) [reply Framework/PythonInterface/core/inc/MantidPythonInterface/core/PythonStdoutChannel.h:49]: I added a "TODO" line stating we should replace the current python channel to the one using pybind11, once this library is adopted in Mantid.

File: Framework/PythonInterface/core/src/PythonStdoutChannel.cpp (hunk 2 above)
Comment (jmborr) [reply Framework/PythonInterface/core/src/PythonStdoutChannel.cpp:42]: I just renamed the variable to explicitly state that this object *is* an object of type `std::stream`

File: Framework/PythonInterface/mantid/kernel/src/kernel.cpp.in (hunk 3 above)
Comment (jmborr) [reply Framework/PythonInterface/mantid/kernel/src/kernel.cpp.in:54]: I replaced with the same TODO comment as before

**Review by wdzhou — DISMISSED** [review]
It works to me.

:shipit:

File: Framework/PythonInterface/core/inc/MantidPythonInterface/core/PythonStdoutChannel.h [hunk 6]
```diff
@@ -27,4 +36,18 @@ class MANTID_PYTHONINTERFACE_CORE_DLL PythonStdoutChannel : public ConsoleChanne
   /// Constructor for PythonStdoutChannel
   PythonStdoutChannel();
 };
+
+/*
+// TODO: this channel should replace PythonStdoutChannel when we adopt pybind11 because of robust GIL management
```
Comment (martyngigg) [inline Framework/PythonInterface/core/inc/MantidPythonInterface/core/PythonStdoutChannel.h:41]: It's not really clear if we will adopt `pybind11` - there is a question whether the port is worth it?

When you say robust GIL management, do you mean automatic?

File: Framework/PythonInterface/core/inc/MantidPythonInterface/core/PythonStdoutChannel.h [hunk 7]
```diff
@@ -17,8 +17,17 @@
...
 #include "MantidPythonInterface/core/DllConfig.h"
+#include "MantidPythonInterface/core/WrapPython.h"
+
+// 3rd-party includes
+#include "MantidKernel/StdoutChannel.h"
 #include <Poco/ConsoleChannel.h>
+//#include <pybind11/iostream.h>
+
+// standard includes
+#include <iostream>
```
Comment (martyngigg) [inline Framework/PythonInterface/core/inc/MantidPythonInterface/core/PythonStdoutChannel.h:30]: Can we remove this from the header? It's very heavy to include and we shouldn't need to include it in a header.

**Review by martyngigg — CHANGES_REQUESTED** [review]

File: Framework/PythonInterface/core/inc/MantidPythonInterface/core/PythonStdoutChannel.h (hunk 6 above)
Comment (jmborr) [reply Framework/PythonInterface/core/inc/MantidPythonInterface/core/PythonStdoutChannel.h:41]: pybind11 class [gil_scoped_acquire](https://github.com/pybind/pybind11/blob/master/include/pybind11/gil.h) is the analogous of our class [GlobalInterpreterLock](https://github.com/mantidproject/mantid/blob/master/Framework/PythonInterface/core/inc/MantidPythonInterface/core/GlobalInterpreterLock.h#L23). This pybind11 class seem to address [a few scenarios](https://github.com/pybind/pybind11/blob/master/include/pybind11/gil.h#L26).

File: Framework/PythonInterface/core/inc/MantidPythonInterface/core/PythonStdoutChannel.h (hunk 7 above)
Comment (jmborr) [reply Framework/PythonInterface/core/inc/MantidPythonInterface/core/PythonStdoutChannel.h:30]: thanks for catching this. It was a leftover from a console channel class I had written before and deleted later. I'll remove these lines

**Review by martyngigg — APPROVED** [review]

**Comment by jmborr** [comment]
@wdzhou @martyngigg 🕺 thank you

**Pull request merged**